File: packages/go/wallet.go
package constellation

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"strings"

	"github.com/btcsuite/btcd/btcec/v2"
)

const base58Alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

// GenerateKeyPair creates a new random key pair
func GenerateKeyPair() (*KeyPair, error) {
	privateKey, err := btcec.NewPrivateKey()
	if err != nil {
		return nil, fmt.Errorf("failed to generate private key: %w", err)
	}

	privateKeyHex := hex.EncodeToString(privateKey.Serialize())
	publicKeyHex := hex.EncodeToString(privateKey.PubKey().SerializeUncompressed())
	address := GetAddress(publicKeyHex)

	return &KeyPair{
		PrivateKey: privateKeyHex,
		PublicKey:  publicKeyHex,
		Address:    address,
	}, nil
}

// KeyPairFromPrivateKey derives a key pair from an existing private key
func KeyPairFromPrivateKey(privateKeyHex string) (*KeyPair, error) {
	if !IsValidPrivateKey(privateKeyHex) {
		return nil, ErrInvalidPrivateKey
	}

	privateKeyBytes, err := hex.DecodeString(privateKeyHex)
	if err != nil {
		return nil, fmt.Errorf("invalid hex string: %w", err)
	}

	privateKey, _ := btcec.PrivKeyFromBytes(privateKeyBytes)
	publicKeyHex := hex.EncodeToString(privateKey.PubKey().SerializeUncompressed())
	address := GetAddress(publicKeyHex)

	return &KeyPair{
		PrivateKey: privateKeyHex,
		PublicKey:  publicKeyHex,
		Address:    address,
	}, nil
}

// GetPublicKeyHex returns the public key hex from a private key
func GetPublicKeyHex(privateKeyHex string, compressed bool) (string, error) {
	privateKeyBytes, err := hex.DecodeString(privateKeyHex)
	if err != nil {
		return "", fmt.Errorf("invalid hex string: %w", err)
	}

	privateKey, _ := btcec.PrivKeyFromBytes(privateKeyBytes)

	if compressed {
		return hex.EncodeToString(privateKey.PubKey().SerializeCompressed()), nil
	}
	return hex.EncodeToString(privateKey.PubKey().SerializeUncompressed()), nil
}

// GetPublicKeyID returns the public key ID (without 04 prefix) from a private key
// This format is used in SignatureProof.ID
func GetPublicKeyID(privateKeyHex string) (string, error) {
	publicKey, err := GetPublicKeyHex(privateKeyHex, false)
	if err != nil {
		return "", err
	}
	return NormalizePublicKeyToID(publicKey), nil
}

// GetAddress derives a DAG address from a public key
func GetAddress(publicKeyHex string) string {
	// PKCS prefix for X.509 DER encoding (secp256k1)
	pkcsPrefix := "3056301006072a8648ce3d020106052b8104000a034200"

	// Normalize public key to include 04 prefix
	normalizedKey := NormalizePublicKey(publicKeyHex)

	// Prepend PKCS prefix
	pkcsEncoded := pkcsPrefix + normalizedKey

	// SHA-256 hash
	pkcsBytes, _ := hex.DecodeString(pkcsEncoded)
	hash := sha256.Sum256(pkcsBytes)

	// Base58 encode
	encoded := base58Encode(hash[:])

	// Take last 36 characters
	last36 := encoded
	if len(encoded) > 36 {
		last36 = encoded[len(encoded)-36:]
	}

	// Calculate parity digit (sum of numeric characters mod 9)
	digitSum := 0
	for _, c := range last36 {
		if c >= '0' && c <= '9' {
			digitSum += int(c - '0')
		}
	}
	parity := digitSum % 9

	// Return with DAG prefix, parity, and last36
	return fmt.Sprintf("DAG%d%s", parity, last36)
}

// IsValidPrivateKey validates that a private key is correctly formatted
func IsValidPrivateKey(privateKeyHex string) bool {
	if len(privateKeyHex) != 64 {
		return false
	}
	for _, c := range privateKeyHex {
		if !isHexChar(c) {
			return false
		}
	}
	return true
}

// IsValidPublicKey validates that a public key is correctly formatted
func IsValidPublicKey(publicKeyHex string) bool {
	// With 04 prefix: 130 chars, without: 128 chars
	if len(publicKeyHex) != 128 && len(publicKeyHex) != 130 {
		return false
	}
	for _, c := range publicKeyHex {
		if !isHexChar(c) {
			return false
		}
	}
	return true
}

// NormalizePublicKey ensures the public key has the 04 prefix
func NormalizePublicKey(publicKeyHex string) string {
	if len(publicKeyHex) == 128 {
		return "04" + publicKeyHex
	}
	return publicKeyHex
}

// NormalizePublicKeyToID returns the public key without the 04 prefix
func NormalizePublicKeyToID(publicKeyHex string) string {
	if len(publicKeyHex) == 130 && strings.HasPrefix(publicKeyHex, "04") {
		return publicKeyHex[2:]
	}
	return publicKeyHex
}

func isHexChar(c rune) bool {
	return (c >= '0' && c <= '9') || (c >= 'a' && c <= 'f') || (c >= 'A' && c <= 'F')
}

// base58Encode encodes bytes using Bitcoin/Constellation alphabet
func base58Encode(data []byte) string {
	if len(data) == 0 {
		return ""
	}

	// Count leading zeros
	leadingZeros := 0
	for _, b := range data {
		if b == 0 {
			leadingZeros++
		} else {
			break
		}
	}

	// Convert to big integer representation
	var digits []byte
	for _, b := range data {
		carry := int(b)
		for i := range digits {
			carry += int(digits[i]) << 8
			digits[i] = byte(carry % 58)
			carry /= 58
		}
		for carry > 0 {
			digits = append(digits, byte(carry%58))
			carry /= 58
		}
	}

	// Build result string
	result := make([]byte, 0, leadingZeros+len(digits))

	// Add '1' for each leading zero byte
	for i := 0; i < leadingZeros; i++ {
		result = append(result, '1')
	}

	// Convert digits to characters (in reverse order)
	for i := len(digits) - 1; i >= 0; i-- {
		result = append(result, base58Alphabet[digits[i]])
	}

	return string(result)
}

File: packages/rust/src/wallet.rs
//! Wallet and Key Management Utilities
//!
//! Functions for generating and managing cryptographic keys.

use rand::rngs::OsRng;
use secp256k1::{PublicKey, Secp256k1, SecretKey};
use sha2::{Digest, Sha256};

use crate::types::{KeyPair, Result, SdkError};

const BASE58_ALPHABET: &[u8] = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";

/// Generate a new random key pair
///
/// # Example
/// ```
/// use constellation_sdk::wallet::generate_key_pair;
///
/// let key_pair = generate_key_pair();
/// println!("Address: {}", key_pair.address);
/// println!("Private key: {}", key_pair.private_key);
/// println!("Public key: {}", key_pair.public_key);
/// ```
pub fn generate_key_pair() -> KeyPair {
    let secp = Secp256k1::new();
    let (secret_key, public_key) = secp.generate_keypair(&mut OsRng);

    let private_key_hex = hex::encode(secret_key.secret_bytes());
    let public_key_hex = hex::encode(public_key.serialize_uncompressed());
    let address = get_address(&public_key_hex);

    KeyPair {
        private_key: private_key_hex,
        public_key: public_key_hex,
        address,
    }
}

/// Derive a key pair from an existing private key
///
/// # Arguments
/// * `private_key` - Private key in hex format (64 characters)
///
/// # Example
/// ```
/// use constellation_sdk::wallet::{generate_key_pair, key_pair_from_private_key};
///
/// let original = generate_key_pair();
/// let derived = key_pair_from_private_key(&original.private_key).unwrap();
/// assert_eq!(original.public_key, derived.public_key);
/// ```
pub fn key_pair_from_private_key(private_key: &str) -> Result<KeyPair> {
    if !is_valid_private_key(private_key) {
        return Err(SdkError::InvalidPrivateKey(
            "Invalid private key format".to_string(),
        ));
    }

    let secp = Secp256k1::new();
    let private_key_bytes = hex::decode(private_key)?;
    let secret_key = SecretKey::from_slice(&private_key_bytes)?;
    let public_key = PublicKey::from_secret_key(&secp, &secret_key);

    let public_key_hex = hex::encode(public_key.serialize_uncompressed());
    let address = get_address(&public_key_hex);

    Ok(KeyPair {
        private_key: private_key.to_string(),
        public_key: public_key_hex,
        address,
    })
}

/// Get the public key hex from a private key
///
/// # Arguments
/// * `private_key` - Private key in hex format
/// * `compressed` - If true, returns compressed public key (33 bytes)
pub fn get_public_key_hex(private_key: &str, compressed: bool) -> Result<String> {
    let private_key_bytes = hex::decode(private_key)?;
    let secp = Secp256k1::new();
    let secret_key = SecretKey::from_slice(&private_key_bytes)?;
    let public_key = PublicKey::from_secret_key(&secp, &secret_key);

    if compressed {
        Ok(hex::encode(public_key.serialize()))
    } else {
        Ok(hex::encode(public_key.serialize_uncompressed()))
    }
}

/// Get the public key ID (without 04 prefix) from a private key
///
/// This format is used in SignatureProof.id
///
/// # Arguments
/// * `private_key` - Private key in hex format
///
/// # Returns
/// Public key ID (128 characters, no 04 prefix)
pub fn get_public_key_id(private_key: &str) -> Result<String> {
    let public_key = get_public_key_hex(private_key, false)?;
    Ok(normalize_public_key_to_id(&public_key))
}

/// Get DAG address from a public key
///
/// Uses Constellation's address derivation:
/// 1. Normalize public key to include 04 prefix
/// 2. Prepend PKCS prefix (X.509 DER encoding header)
/// 3. SHA-256 hash
/// 4. Base58 encode
/// 5. Take last 36 characters
/// 6. Calculate parity digit (sum of numeric characters mod 9)
/// 7. Result: DAG + parity + last36
///
/// # Arguments
/// * `public_key` - Public key in hex format (with or without 04 prefix)
pub fn get_address(public_key: &str) -> String {
    // PKCS prefix for X.509 DER encoding (secp256k1)
    const PKCS_PREFIX: &str = "3056301006072a8648ce3d020106052b8104000a034200";

    // Normalize public key to include 04 prefix
    let normalized_key = normalize_public_key(public_key);

    // Prepend PKCS prefix
    let pkcs_encoded = format!("{}{}", PKCS_PREFIX, normalized_key);

    // SHA-256 hash
    let pkcs_bytes = hex::decode(&pkcs_encoded).unwrap_or_default();
    let mut hasher = Sha256::new();
    hasher.update(&pkcs_bytes);
    let hash = hasher.finalize();

    // Base58 encode
    let encoded = base58_encode(&hash);

    // Take last 36 characters
    let last36 = if encoded.len() > 36 {
        &encoded[encoded.len() - 36..]
    } else {
        &encoded
    };

    // Calculate parity digit (sum of numeric characters mod 9)
    let digit_sum: u32 = last36
        .chars()
        .filter(|c| c.is_ascii_digit())
        .map(|c| c.to_digit(10).unwrap_or(0))
        .sum();
    let parity = digit_sum % 9;

    // Return with DAG prefix, parity, and last36
    format!("DAG{}{}", parity, last36)
}

/// Validate that a private key is correctly formatted
///
/// # Arguments
/// * `private_key` - Private key to validate
///
/// # Returns
/// true if valid hex string of correct length
pub fn is_valid_private_key(private_key: &str) -> bool {
    if private_key.len() != 64 {
        return false;
    }
    private_key.chars().all(|c| c.is_ascii_hexdigit())
}

/// Validate that a public key is correctly formatted
///
/// # Arguments
/// * `public_key` - Public key to validate
///
/// # Returns
/// true if valid hex string of correct length
pub fn is_valid_public_key(public_key: &str) -> bool {
    // With 04 prefix: 130 chars, without: 128 chars
    if public_key.len() != 128 && public_key.len() != 130 {
        return false;
    }
    public_key.chars().all(|c| c.is_ascii_hexdigit())
}

/// Normalize public key to include 04 prefix
pub fn normalize_public_key(public_key: &str) -> String {
    if public_key.len() == 128 {
        format!("04{}", public_key)
    } else {
        public_key.to_string()
    }
}

/// Normalize public key to ID format (without 04 prefix)
pub fn normalize_public_key_to_id(public_key: &str) -> String {
    if public_key.len() == 130 && public_key.starts_with("04") {
        public_key[2..].to_string()
    } else {
        public_key.to_string()
    }
}

/// Base58 encode bytes using Bitcoin/Constellation alphabet
fn base58_encode(data: &[u8]) -> String {
    if data.is_empty() {
        return String::new();
    }

    // Count leading zeros
    let leading_zeros = data.iter().take_while(|&&b| b == 0).count();

    // Convert to big integer representation
    let mut num: Vec<u8> = Vec::with_capacity(data.len() * 138 / 100 + 1);
    for &byte in data {
        let mut carry = byte as u32;
        for digit in num.iter_mut() {
            carry += (*digit as u32) << 8;
            *digit = (carry % 58) as u8;
            carry /= 58;
        }
        while carry > 0 {
            num.push((carry % 58) as u8);
            carry /= 58;
        }
    }

    // Build result string
    let mut result = String::with_capacity(leading_zeros + num.len());

    // Add '1' for each leading zero byte
    for _ in 0..leading_zeros {
        result.push('1');
    }

    // Convert digits to characters
    for &digit in num.iter().rev() {
        result.push(BASE58_ALPHABET[digit as usize] as char);
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_key_pair() {
        let key_pair = generate_key_pair();
        assert_eq!(key_pair.private_key.len(), 64);
        assert_eq!(key_pair.public_key.len(), 130);
        assert!(key_pair.address.starts_with("DAG"));
    }

    #[test]
    fn test_key_pair_from_private_key() {
        let key_pair = generate_key_pair();
        let derived = key_pair_from_private_key(&key_pair.private_key).unwrap();
        assert_eq!(derived.public_key, key_pair.public_key);
        assert_eq!(derived.address, key_pair.address);
    }

    #[test]
    fn test_is_valid_private_key() {
        assert!(is_valid_private_key(&"a".repeat(64)));
        assert!(!is_valid_private_key(&"a".repeat(63)));
        assert!(!is_valid_private_key(&"g".repeat(64)));
    }

    #[test]
    fn test_is_valid_public_key() {
        assert!(is_valid_public_key(&"a".repeat(128)));
        assert!(is_valid_public_key(&"a".repeat(130)));
        assert!(!is_valid_public_key(&"a".repeat(127)));
    }
}
